src/workerd/io/io-context.h

```diff
@@ -1257,23 +1257,40 @@ kj::_::ReducePromises<RemoveIoOwn<T>> IoContext::awaitJs(jsg::Lock& js, jsg::Pro
   auto paf = kj::newPromiseAndFulfiller<RemoveIoOwn<T>>();
   struct RefcountedFulfiller: public Finalizeable, public kj::Refcounted {
     kj::Own<kj::PromiseFulfiller<RemoveIoOwn<T>>> fulfiller;
+    kj::Own<const workerd::Worker::Isolate> isolate;
     bool isDone = false;
 
-    RefcountedFulfiller(kj::Own<kj::PromiseFulfiller<RemoveIoOwn<T>>> fulfiller)
-        : fulfiller(kj::mv(fulfiller)) {}
+    RefcountedFulfiller(kj::Own<const workerd::Worker::Isolate> isolate,
+        kj::Own<kj::PromiseFulfiller<RemoveIoOwn<T>>> fulfiller)
+        : fulfiller(kj::mv(fulfiller)),
+          isolate(kj::mv(isolate)) {}
 
     ~RefcountedFulfiller() noexcept(false) {
       if (!isDone) {
-        // The JavaScript resolver was garbage collected, i.e. JavaScript will never resolve
-        // this promise.
-        fulfiller->reject(JSG_KJ_EXCEPTION(FAILED, Error, "Promise will never complete."));
+        auto hasExcessivelyExceededHeapLimit =
+            isolate->getLimitEnforcer().hasExcessivelyExceededHeapLimit();
+        if (hasExcessivelyExceededHeapLimit) {
+          fulfiller->reject(
+              JSG_KJ_EXCEPTION(OVERLOADED, Error, "jsg.Error: Worker has exceeded memory limit."));
```

Comment:
Not something to solve for now but this does make me wonder if there are other possible similar cases where we know why the promise will never complete... something to look at later.

Reply:
Hmm, it doesn't seem simple to find similar cases, but maybe I'm missing something.

Reply:
Yeah we'd have to investigate to see. As I said, not something to solve for now but something we may want to explore as a later step